src/connector/src/sink/pulsar.rs

```diff
@@ -0,0 +1,321 @@
+        event_object: Option<serde_json::Value>,
+    ) -> Result<()> {
+        let message = Message {
+            partition_key: event_key_object.map(|key| key.to_string()),
+            payload: serde_json::to_vec(&event_object)
+                .map_err(|e| SinkError::Pulsar(anyhow!(e)))?,
+            ..Default::default()
+        };
+
+        self.send_message(message).await?;
+        Ok(())
+    }
+
+    async fn append_only(&mut self, chunk: StreamChunk) -> Result<()> {
+        // TODO: Remove the clones here, only to satisfy borrow checker at present
+        let schema = self.schema.clone();
+        let downstream_pk = self.downstream_pk.clone();
+        let key_encoder =
+            JsonEncoder::new(&schema, Some(&downstream_pk), TimestampHandlingMode::Milli);
+        let val_encoder = JsonEncoder::new(&schema, None, TimestampHandlingMode::Milli);
+
+        let append_only_stream = gen_append_only_message_stream(
+            chunk,
+            AppendOnlyAdapterOpts::default(),
+            key_encoder,
+            val_encoder,
+        );
+
+        #[for_await]
+        for msg in append_only_stream {
+            let (event_key_object, event_object) = msg?;
+            self.write_json_objects(event_key_object, event_object).await?;
+        }
+        Ok(())
+    }
+
+    async fn upsert(&mut self, chunk: StreamChunk) -> Result<()> {
+        // TODO: Remove the clones here, only to satisfy borrow checker at present
+        let schema = self.schema.clone();
+        let downstream_pk = self.downstream_pk.clone();
+        let key_encoder =
+            JsonEncoder::new(&schema, Some(&downstream_pk), TimestampHandlingMode::Milli);
+        let val_encoder = JsonEncoder::new(&schema, None, TimestampHandlingMode::Milli);
+
+        let upsert_stream = gen_upsert_message_stream(
+            chunk, 
+            UpsertAdapterOpts::default(),
+            key_encoder,
+            val_encoder,
+        );
+
+        #[for_await]
+        for msg in upsert_stream {
+            let (event_key_object, event_object) = msg?;
+            self.write_json_objects(event_key_object, event_object).await?;
+        }
+        Ok(())
+    }
+
+    async fn commit_inner(&mut self) -> Result<()> {
+        try_join_all(
+            self.send_future_buffer
+                .drain(..)
+                .map(|send_future| send_future
+                    .map_err(|e| SinkError::Pulsar(anyhow!(e)))
+                ),
+        ).await?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl SinkWriter for PulsarSinkWriter {
+    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+        if self.is_append_only {
+            self.append_only(chunk).await
+        } else {
+            self.upsert(chunk).await
+        }
+    }
+
+    async fn begin_epoch(&mut self, _epoch: u64) -> Result<()> {
+        Ok(())
+    }
+
+    async fn barrier(&mut self, is_checkpoint: bool) -> Result<Self::CommitMetadata> {
```

Comment:
In `pulsar` crate we have a `send_batch` function, to ensure all batched messages are sent.
Should we add a `send_batch` before `commit_inner`, in this case we do not need to `await` unnecessary futures, cc @tabVersion.
```rust
/// sends the current batch of messages
#[cfg_attr(feature = "telemetry", tracing::instrument(skip_all))]
pub async fn send_batch(&mut self) -> Result<(), Error> {
    match &mut self.inner {
        ProducerInner::Single(p) => p.send_batch().await,
        ProducerInner::Partitioned(p) => {
            try_join_all(p.producers.iter_mut().map(|p| p.send_batch()))
                .await
                .map(drop)
        }
    }
}
```

Reply:
In my understanding, send_batch ensures the messages in the current batch are sent, but does not ensure the receipt of the message held by the SendFuture (returned by producer.send). If messages are sent in batches, indeed we need to call send_batch first during commit. But we still need to await on all SendFutures to ensure all receipts.

Reply:
This is to accelerate the later `await` speed, since we already ensure the receipt of the messages from this `send_batch`. So I think it's good to add one before the actual group await.